Switch to Go router.

kormat opened the pull request:
- This change adds SCMP error handling for the Go router. Any error
  that's raised that corresponds to an SCMP error has the SCMP class,
  type, and any required info, embedded in it, allowing for generic
  handling and processing.
- SIBRA integration tests are no longer run by default, as the Go router
  does not yet have SIBRA support.

Major refactorings:
- Move `RawBytes` and `Error` from go/lib/util to go/lib/common, as
  they're used in ~all go code.
- Rename go/border/packet.packet to go/border/rpkt.RtrPkt (for Router
  Packet), and add go/lib/spkt.ScnPkt (for SCION Packet), and functions
  for converting between them.
  - RtrPkt is a high-performance representation that's tied to a single
    underlying byte slice. It only parses exactly the bytes required, on
    demand. As such it only allows very minimal, in-place, packet
    modification.
  - ScnPkt is a high-level representation that has no underlying storage,
    allowing it to be easily maniuplated (e.g. reversing,
    adding/changing/removing extensions, etc).
- Split up go/border/rpkt/extn_* into router-specific logic (stays in
  the original dir) and non-router-specific logic in go/lib/spkt/extn_*.
  The router specific types have an `R` prefix to help distinguish them.
- Defined interfaces for extensions, l4 headers, and payloads, so they
  can be handled generically.

Also:
- Replace scion.sh's gofmt and govet commands with a unified golint
  command.
- Update govendor now that
  kardianos/govendor#242 is merged.
- Added go/lib/assert to make it easier to check prereqs/assumptions in
  code. It's turned on by default, and can be disabled via
  `make GOTAGS= all` in `go/`.
- Add goimports tool as a dependency, to find formatting issues with
  imports.
- Improve goconvey console text output by using the `story` reporter.
- Store raw RevInfo messages in the go router for use in SCMP revocation
  messages.
- Set a maximum virtual memory size for the go router, to help catch
  run-away memory leaks before the machine starts swap thrashing.
- Create all RtrPkt's with maximum-sized buffers, which means they can all
  be reused.
- Moved all the binary `order` variables to a single entry in
  go/lib/common
- Add scmp revocation handling to the path server.
- Add some tests to scmp_error_test.py that don't expect a reply, just
  to make sure the router doesn't crash because of them.
- Only run goimport on local non-generated files, to avoid errors from
  go/proto/*.capnp.go.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/netsec-ethz/scion/934)
<!-- Reviewable:end -->

kormat commented:
@klausman 
Notes:
- I'm not finished implementing all the error codes yet, hence a bunch of tests in scmp_error_test are currently commented out.
- The code needs a lot more commenting, so please point out places where comments are desired.

klausman commented:
One early comment before I dive deeper.

---

Reviewed 4 of 75 files at r1.
Review status: 4 of 74 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[go/border/error.go, line 1 at r1](https://reviewable.io:443/reviews/netsec-ethz/scion/934#-KWhJGk-DhKzxiw3WbNR:-KWhJGk-DhKzxiw3WbNS:bwtldzw) ([raw file](https://github.com/netsec-ethz/scion/blob/d714ccca34bbfe8b0b0722a2d8078d5f7ee1e087/go/border/error.go#L1)):*
> ```Go
> // Copyright 2016 ETH Zurich
> ```

One toplevel/general comment: some of the variable names are really hard to read, in the sense of being opaque:

What does the s in spkt stand for? Source? SCMP? Something else? In that sense, scmpp would be more useful: the interesting part is the SCMP bit, not that it's a packet.
Similar for rpkt and maybe a few others.

It gets even more confusing with "rp" -- is this a path? a packet? and what does the r stand for?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion/934#-:-KWhK3nCcAELLJqCk-aD:bf30e9w)*
<!-- Sent from Reviewable.io -->

kormat commented:
Review status: 3 of 75 files reviewed at latest revision, 1 unresolved discussion.

---

*[go/border/error.go, line 1 at r1](https://reviewable.io:443/reviews/netsec-ethz/scion/934#-KWhJGk-DhKzxiw3WbNR:-KWhbpRyfCmURAdrT6cp:b-px4aqg) ([raw file](https://github.com/netsec-ethz/scion/blob/d714ccca34bbfe8b0b0722a2d8078d5f7ee1e087/go/border/error.go#L1)):*
<details><summary><i>Previously, klausman (Tobias Klausmann) wrote…</i></summary>
> One toplevel/general comment: some of the variable names are really hard to read, in the sense of being opaque:
> 
> What does the s in spkt stand for? Source? SCMP? Something else? In that sense, scmpp would be more useful: the interesting part is the SCMP bit, not that it's a packet.
> Similar for rpkt and maybe a few others.
> 
> It gets even more confusing with "rp" -- is this a path? a packet? and what does the r stand for?
</details>
Renamed the types as discussed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion/934)*
<!-- Sent from Reviewable.io -->

klausman commented:
Reviewed 3 of 75 files at r1, 18 of 27 files at r2.
Review status: 24 of 75 files reviewed at latest revision, 11 unresolved discussions.

---

*[go/border/main.go, line 47 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/934#-KWhRIfyZHfpCwvaOOAt:-KWhRIfyZHfpCwvaOOAu:bgk2ov4) ([raw file](https://github.com/netsec-ethz/scion/blob/7b7cc3ade0ffb1f1ffce7bf0b3545f53aeb88aa4/go/border/main.go#L47)):*
> ```Go
> 	}
> 	setupSignals()
> 	// Set max virtual memory size to 1G in bytes.
> ```

Change the end to "1GB" or "1GiB"

---

*[go/border/hsr/hsr.go, line 121 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/934#-KWhT8a42jhF0q4DvgLL:-KWhT8a42jhF0q4DvgLM:bv8nu6t) ([raw file](https://github.com/netsec-ethz/scion/blob/7b7cc3ade0ffb1f1ffce7bf0b3545f53aeb88aa4/go/border/hsr/hsr.go#L121)):*
> ```Go
> 			return nil, err
> 		}
> 		rp.Ingress.Dst = AddrMs[cp.port_id].GoAddr
> ```

What is AddrMs and where does it come from? Maybe a comment to that effect would be good.

---

*[go/border/rpkt/create.go, line 84 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/934#-KWhXEDqO9drFmEWeMZq:-KWhXEDrLti0WGIIjT5_:b-766w16) ([raw file](https://github.com/netsec-ethz/scion/blob/7b7cc3ade0ffb1f1ffce7bf0b3545f53aeb88aa4/go/border/rpkt/create.go#L84)):*
> ```Go
> 		}
> 	} else {
> 		// Trim buffer
> ```

Expand comment to explain what the buffer is trimmed to (and why).

---

*[go/border/rpkt/create.go, line 103 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/934#-KWhXsdmYuPX38Cz5pl8:-KWhXsdmYuPX38Cz5pl9:b-odex49) ([raw file](https://github.com/netsec-ethz/scion/blob/7b7cc3ade0ffb1f1ffce7bf0b3545f53aeb88aa4/go/border/rpkt/create.go#L103)):*
> ```Go
> 	rp.idxs.pld = rp.idxs.l4 + l4h.L4Len()
> 	var nextHdr *uint8
> 	for _, eIdx := range rp.idxs.hbhExt {
> ```

Add comment here (and to the loop below) that expands HBH/E2E to hop-by-hop and end-to-end.

---

*[go/border/rpkt/create.go, line 128 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/934#-KWhYNaweUAGIFqoWeoe:-KWhYNaweUAGIFqoWeof:btwec4v) ([raw file](https://github.com/netsec-ethz/scion/blob/7b7cc3ade0ffb1f1ffce7bf0b3545f53aeb88aa4/go/border/rpkt/create.go#L128)):*
> ```Go
> 		// Reset buffer to full size
> 		rp.Raw = rp.Raw[:cap(rp.Raw)-1]
> 		// Write out payload
> ```

Stylistic note: instead of "Write out X" here and elsewhere, use "Write X into Y" (where Y here is the payload buffer), as it is more informative and doesn't connote "write to stdout/err" or "...to the network" ("out" sort of implying "outside of the program's memory")

---

*[go/border/rpkt/create.go, line 135 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/934#-KWhYqI9hEhqFxynkn13:-KWhYqI9hEhqFxynkn14:bnyok86) ([raw file](https://github.com/netsec-ethz/scion/blob/7b7cc3ade0ffb1f1ffce7bf0b3545f53aeb88aa4/go/border/rpkt/create.go#L135)):*
> ```Go
> 		}
> 	}
> 	// Trim buffer
> ```

Again, trim to what? "minimum required size"?

---

*[go/border/rpkt/create.go, line 137 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/934#-KWhYxqPXeKEu7-1cPx6:-KWhYxqPXeKEu7-1cPx7:b-sus84a) ([raw file](https://github.com/netsec-ethz/scion/blob/7b7cc3ade0ffb1f1ffce7bf0b3545f53aeb88aa4/go/border/rpkt/create.go#L137)):*
> ```Go
> 	// Trim buffer
> 	rp.Raw = rp.Raw[:rp.idxs.pld+plen]
> 	// Now start updating headers
> ```

s/Now start updating/Update/

---

*[go/border/rpkt/extns.go, line 77 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/934#-KWhczIILHCs6KOLeZAy:-KWhczIILHCs6KOLeZAz:b-wmfb8l) ([raw file](https://github.com/netsec-ethz/scion/blob/7b7cc3ade0ffb1f1ffce7bf0b3545f53aeb88aa4/go/border/rpkt/extns.go#L77)):*
> ```Go
> 	et := e.Type()
> 	// Set the preceeding NextHdr field, whether it's in the common header, or
> 	// a preceeding hop-by-hopb extension.
> ```

s/hopb/hop/

---

*[go/border/rpkt/extns.go, line 79 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/934#-KWhd36mz3t2qln8SBv2:-KWhd36mz3t2qln8SBv3:bd7r9z2) ([raw file](https://github.com/netsec-ethz/scion/blob/7b7cc3ade0ffb1f1ffce7bf0b3545f53aeb88aa4/go/border/rpkt/extns.go#L79)):*
> ```Go
> 	// a preceeding hop-by-hopb extension.
> 	*nextHdr = uint8(et.Class)
> 	// Check the extension's length is legal
> ```

better: // Check if the extenion's length is correct

(or legal, if you want to discern external vs. internal consistency; but do add the "if")

---

*[go/border/rpkt/parse.go, line 147 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/934#-KWhfpWLic37ca864dj7:-KWhfpWLic37ca864dj8:bura2ht) ([raw file](https://github.com/netsec-ethz/scion/blob/7b7cc3ade0ffb1f1ffce7bf0b3545f53aeb88aa4/go/border/rpkt/parse.go#L147)):*
> ```Go
> 	if *offset > len(rp.Raw) {
> 		// Can't generate SCMP error in general as we can't parse anything
> 		// after the hbh extensions (i.e. the l4 header).
> ```

Should this i.e. be an e.g.?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion/934)*
<!-- Sent from Reviewable.io -->

klausman commented:
Reviewed 28 of 75 files at r1, 8 of 27 files at r2.
Review status: 59 of 75 files reviewed at latest revision, 14 unresolved discussions.

---

*[go/lib/scmp/extn.go, line 87 at r3](https://reviewable.io:443/reviews/netsec-ethz/scion/934#-KWhkZOnJVbQuhcuF2DO:-KWhkZOnJVbQuhcuF2DP:bus5pox) ([raw file](https://github.com/netsec-ethz/scion/blob/7d2c380b675e3bf988e500e41d2ad911abd5c9a7/go/lib/scmp/extn.go#L87)):*
> ```Go
> 
> func (e *Extn) String() string {
> 	return fmt.Sprintf("SCMP Ext(%dB): Error? %v HopByHop: %v", e.Len(), e.Error, e.HopByHop)
> ```

Should the ? after Error be a : ?

---

*[go/lib/scmp/info.go, line 178 at r3](https://reviewable.io:443/reviews/netsec-ethz/scion/934#-KWhmN_hJ4o6X8nkpavs:-KWhmN_hJ4o6X8nkpavt:bbhcdol) ([raw file](https://github.com/netsec-ethz/scion/blob/7d2c380b675e3bf988e500e41d2ad911abd5c9a7/go/lib/scmp/info.go#L178)):*
> ```Go
> 	revToken common.RawBytes) *InfoRevocation {
> 	return &InfoRevocation{
> 		InfoPathOffsets: &InfoPathOffsets{InfoF: infoF, HopF: hopF, IfID: IfID, Ingress: ingress},
> ```

All RHS start with a lower-case letter, except `IfID`

---

*[go/lib/scmp/info.go, line 233 at r3](https://reviewable.io:443/reviews/netsec-ethz/scion/934#-KWhml5BNx5z_ntPHh5A:-KWhml5C9yMglsBHIcm5:be4r0g0) ([raw file](https://github.com/netsec-ethz/scion/blob/7d2c380b675e3bf988e500e41d2ad911abd5c9a7/go/lib/scmp/info.go#L233)):*
> ```Go
> func (r *InfoExtIdx) Len() int {
> 	l := 1
> 	return l + util.CalcPadding(l, common.LineLen)
> ```

couldn't `l` be dropped here and just replaced with `1` in the two spots it's used?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion/934)*
<!-- Sent from Reviewable.io -->

kormat commented:
Review status: 59 of 75 files reviewed at latest revision, 14 unresolved discussions.

---

*[go/border/main.go, line 47 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/934#-KWhRIfyZHfpCwvaOOAt:-KWhgL-E5BI7LM8BzGZG:b-d7ac0f) ([raw file](https://github.com/netsec-ethz/scion/blob/7b7cc3ade0ffb1f1ffce7bf0b3545f53aeb88aa4/go/border/main.go#L47)):*
<details><summary><i>Previously, klausman (Tobias Klausmann) wrote…</i></summary>
> Change the end to "1GB" or "1GiB"
</details>
The `in bytes` part is still important though, as it specifies the unit. I can change it to `1GiB in bytes`.

---

*[go/border/hsr/hsr.go, line 121 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/934#-KWhT8a42jhF0q4DvgLL:-KWhhXgiLpTlICJ3Rw8l:b-wztoe5) ([raw file](https://github.com/netsec-ethz/scion/blob/7b7cc3ade0ffb1f1ffce7bf0b3545f53aeb88aa4/go/border/hsr/hsr.go#L121)):*
<details><summary><i>Previously, klausman (Tobias Klausmann) wrote…</i></summary>
> What is AddrMs and where does it come from? Maybe a comment to that effect would be good.
</details>
Added some comments to the `AddrMeta` type, `AddrMs` variable, and `Init`.

---

*[go/border/rpkt/create.go, line 84 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/934#-KWhXEDqO9drFmEWeMZq:-KWhhsBooG3WZnVPJupX:b-896fix) ([raw file](https://github.com/netsec-ethz/scion/blob/7b7cc3ade0ffb1f1ffce7bf0b3545f53aeb88aa4/go/border/rpkt/create.go#L84)):*
<details><summary><i>Previously, klausman (Tobias Klausmann) wrote…</i></summary>
> Expand comment to explain what the buffer is trimmed to (and why).
</details>
Done.

---

*[go/border/rpkt/create.go, line 128 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/934#-KWhYNaweUAGIFqoWeoe:-KWhoZUSsuxMO-UwUnKA:b-896fix) ([raw file](https://github.com/netsec-ethz/scion/blob/7b7cc3ade0ffb1f1ffce7bf0b3545f53aeb88aa4/go/border/rpkt/create.go#L128)):*
<details><summary><i>Previously, klausman (Tobias Klausmann) wrote…</i></summary>
> Stylistic note: instead of "Write out X" here and elsewhere, use "Write X into Y" (where Y here is the payload buffer), as it is more informative and doesn't connote "write to stdout/err" or "...to the network" ("out" sort of implying "outside of the program's memory")
</details>
Done.

---

*[go/border/rpkt/create.go, line 135 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/934#-KWhYqI9hEhqFxynkn13:-KWhp1EV_hBmsx6MrHJk:b-896fix) ([raw file](https://github.com/netsec-ethz/scion/blob/7b7cc3ade0ffb1f1ffce7bf0b3545f53aeb88aa4/go/border/rpkt/create.go#L135)):*
<details><summary><i>Previously, klausman (Tobias Klausmann) wrote…</i></summary>
> Again, trim to what? "minimum required size"?
</details>
Done.

---

*[go/border/rpkt/create.go, line 137 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/934#-KWhYxqPXeKEu7-1cPx6:-KWhp9W9oUcPjSZUYvUR:b-896fix) ([raw file](https://github.com/netsec-ethz/scion/blob/7b7cc3ade0ffb1f1ffce7bf0b3545f53aeb88aa4/go/border/rpkt/create.go#L137)):*
<details><summary><i>Previously, klausman (Tobias Klausmann) wrote…</i></summary>
> s/Now start updating/Update/
</details>
Done.

---

*[go/border/rpkt/extns.go, line 77 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/934#-KWhczIILHCs6KOLeZAy:-KWhpDGHcNygg2gL551i:b-896fix) ([raw file](https://github.com/netsec-ethz/scion/blob/7b7cc3ade0ffb1f1ffce7bf0b3545f53aeb88aa4/go/border/rpkt/extns.go#L77)):*
<details><summary><i>Previously, klausman (Tobias Klausmann) wrote…</i></summary>
> s/hopb/hop/
</details>
Done.

---

*[go/border/rpkt/extns.go, line 79 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/934#-KWhd36mz3t2qln8SBv2:-KWhpLCqNr9rNM2JbH5c:bwu2q1q) ([raw file](https://github.com/netsec-ethz/scion/blob/7b7cc3ade0ffb1f1ffce7bf0b3545f53aeb88aa4/go/border/rpkt/extns.go#L79)):*
<details><summary><i>Previously, klausman (Tobias Klausmann) wrote…</i></summary>
> better: // Check if the extenion's length is correct
> 
> (or legal, if you want to discern external vs. internal consistency; but do add the "if")
</details>
"Correct" can't be judged, but "legal" can. Added the "if".

---

*[go/border/rpkt/parse.go, line 147 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/934#-KWhfpWLic37ca864dj7:-KWhpxJLTPfPobLyoyDt:b-896fix) ([raw file](https://github.com/netsec-ethz/scion/blob/7b7cc3ade0ffb1f1ffce7bf0b3545f53aeb88aa4/go/border/rpkt/parse.go#L147)):*
<details><summary><i>Previously, klausman (Tobias Klausmann) wrote…</i></summary>
> Should this i.e. be an e.g.?
</details>
Done.

---

*[go/lib/scmp/extn.go, line 87 at r3](https://reviewable.io:443/reviews/netsec-ethz/scion/934#-KWhkZOnJVbQuhcuF2DO:-KWhq-S7kZhV78sjTW3_:b2nhky0) ([raw file](https://github.com/netsec-ethz/scion/blob/7d2c380b675e3bf988e500e41d2ad911abd5c9a7/go/lib/scmp/extn.go#L87)):*
<details><summary><i>Previously, klausman (Tobias Klausmann) wrote…</i></summary>
> Should the ? after Error be a : ?
</details>
No.

---

*[go/lib/scmp/info.go, line 178 at r3](https://reviewable.io:443/reviews/netsec-ethz/scion/934#-KWhmN_hJ4o6X8nkpavs:-KWhqLh7nlG6HbapFKOL:b-896fix) ([raw file](https://github.com/netsec-ethz/scion/blob/7d2c380b675e3bf988e500e41d2ad911abd5c9a7/go/lib/scmp/info.go#L178)):*
<details><summary><i>Previously, klausman (Tobias Klausmann) wrote…</i></summary>
> All RHS start with a lower-case letter, except `IfID`
</details>
Done.

---

*[go/lib/scmp/info.go, line 233 at r3](https://reviewable.io:443/reviews/netsec-ethz/scion/934#-KWhml5BNx5z_ntPHh5A:-KWhqMa6z5R7k2CwwgIv:b-896fix) ([raw file](https://github.com/netsec-ethz/scion/blob/7d2c380b675e3bf988e500e41d2ad911abd5c9a7/go/lib/scmp/info.go#L233)):*
<details><summary><i>Previously, klausman (Tobias Klausmann) wrote…</i></summary>
> couldn't `l` be dropped here and just replaced with `1` in the two spots it's used?
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion/934)*
<!-- Sent from Reviewable.io -->

kormat commented:
Review status: 52 of 75 files reviewed at latest revision, 14 unresolved discussions.

---

*[go/border/rpkt/create.go, line 103 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/934#-KWhXsdmYuPX38Cz5pl8:-KWhqmrhX-Hm8FTNG8nX:b-896fix) ([raw file](https://github.com/netsec-ethz/scion/blob/7b7cc3ade0ffb1f1ffce7bf0b3545f53aeb88aa4/go/border/rpkt/create.go#L103)):*
<details><summary><i>Previously, klausman (Tobias Klausmann) wrote…</i></summary>
> Add comment here (and to the loop below) that expands HBH/E2E to hop-by-hop and end-to-end.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion/934)*
<!-- Sent from Reviewable.io -->

klausman commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> -- Regarding comments, I'll make a separate pass on the files themselves and let you know

---

Reviewed 14 of 75 files at r1, 1 of 27 files at r2, 1 of 1 files at r3, 11 of 11 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion/934#-:-KWlZCkOvmvAoU9URgsw:bo2wysy)*
<!-- Sent from Reviewable.io -->

klausman commented:
One typo in a comment, still <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 10 of 10 files at r5.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[go/border/rpkt/path.go, line 78 at r5](https://reviewable.io:443/reviews/netsec-ethz/scion/934#-KWqYwOkdqwv6vk247Vq:-KWqYwOl3BlUhflE3N_W:b-1li22f) ([raw file](https://github.com/netsec-ethz/scion/blob/801b5f7ba51eb9f6a7f7b23c3d231d86e7236481/go/border/rpkt/path.go#L78)):*
> ```Go
> 	if !ok || info.P.Active() || rp.DirTo == DirSelf {
> 		// Either the interface isn't revoked, or the packet is to this
> 		// router, in which case revokations are ignored to allow communication
> ```

s/revok/revoc/

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion/934#-:-KWqZhOonixaJ0wX6d-7:b-hiqohy)*
<!-- Sent from Reviewable.io -->

klausman commented:
Reviewed 20 of 20 files at r6.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[go/border/rpkt/extns.go, line 68 at r6](https://reviewable.io:443/reviews/netsec-ethz/scion/934#-KWrrQUkeq7v0LmeZB_h:-KWrrQUkeq7v0LmeZB_i:b-gvp7e1) ([raw file](https://github.com/netsec-ethz/scion/blob/b8dabdb42360c05eb94e4db9653749eeb0dfd3ce/go/border/rpkt/extns.go#L68)):*
> ```Go
> 	}
> 	if len(rp.HBHExt) > 1 && e.Type() == common.ExtnSCMPType {
> 		return common.NewError("Bad extension order - SCMP must be first", "idx", len(rp.HBHExt))
> ```

Does this log what the code encountered instead of SCMP? If not, it should.

---

*[go/border/rpkt/extns.go, line 84 at r6](https://reviewable.io:443/reviews/netsec-ethz/scion/934#-KWrrd1KFu8kuDvqBmGT:-KWrrd1KFu8kuDvqBmGU:bprnpm2) ([raw file](https://github.com/netsec-ethz/scion/blob/b8dabdb42360c05eb94e4db9653749eeb0dfd3ce/go/border/rpkt/extns.go#L84)):*
> ```Go
> 	et := e.Type()
> 	// Set the preceeding NextHdr field, whether it's in the common header, or
> 	// a preceeding hop-by-hop extension.
> ```

preceeding -> preceding (2x)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion/934)*
<!-- Sent from Reviewable.io -->

kormat commented:
Review status: 74 of 77 files reviewed at latest revision, 3 unresolved discussions.

---

*[go/border/rpkt/extns.go, line 68 at r6](https://reviewable.io:443/reviews/netsec-ethz/scion/934#-KWrrQUkeq7v0LmeZB_h:-KWs2RxxCfXzVcPjiARS:b-896fix) ([raw file](https://github.com/netsec-ethz/scion/blob/b8dabdb42360c05eb94e4db9653749eeb0dfd3ce/go/border/rpkt/extns.go#L68)):*
<details><summary><i>Previously, klausman (Tobias Klausmann) wrote…</i></summary>
> Does this log what the code encountered instead of SCMP? If not, it should.
</details>
Done.

---

*[go/border/rpkt/extns.go, line 84 at r6](https://reviewable.io:443/reviews/netsec-ethz/scion/934#-KWrrd1KFu8kuDvqBmGT:-KWs315zQrOF4ncFP6JO:b-896fix) ([raw file](https://github.com/netsec-ethz/scion/blob/b8dabdb42360c05eb94e4db9653749eeb0dfd3ce/go/border/rpkt/extns.go#L84)):*
<details><summary><i>Previously, klausman (Tobias Klausmann) wrote…</i></summary>
> preceeding -> preceding (2x)
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion/934)*
<!-- Sent from Reviewable.io -->

kormat commented:
Review status: 74 of 77 files reviewed at latest revision, 3 unresolved discussions.

---

*[go/border/rpkt/path.go, line 78 at r5](https://reviewable.io:443/reviews/netsec-ethz/scion/934#-KWqYwOkdqwv6vk247Vq:-KWs36MR_LQTCEUdeW_v:b-896fix) ([raw file](https://github.com/netsec-ethz/scion/blob/801b5f7ba51eb9f6a7f7b23c3d231d86e7236481/go/border/rpkt/path.go#L78)):*
<details><summary><i>Previously, klausman (Tobias Klausmann) wrote…</i></summary>
> s/revok/revoc/
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion/934)*
<!-- Sent from Reviewable.io -->

klausman commented:
Reviewed 3 of 3 files at r7.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[go/border/rpkt/rpkt.go, line 60 at r7](https://reviewable.io:443/reviews/netsec-ethz/scion/934#-KWs3W15UyHI-BIybnyV:-KWs3W15UyHI-BIybnyW:b-rlqc8u) ([raw file](https://github.com/netsec-ethz/scion/blob/e3dadbea96e3d342d8d916eb7d3354d175e94da2/go/border/rpkt/rpkt.go#L60)):*
> ```Go
> 	// logging statements. (RECV)
> 	Id string
> 	// Raw it the underlying buffer that represents the raw packet bytes. (RECV)
> ```

s/it/is/

---

*[go/border/rpkt/rpkt.go, line 73 at r7](https://reviewable.io:443/reviews/netsec-ethz/scion/934#-KWs3f4Gcl645kRvi6hV:-KWs3f4HDLFRZU2gg9Z6:b-w0drju) ([raw file](https://github.com/netsec-ethz/scion/blob/e3dadbea96e3d342d8d916eb7d3354d175e94da2/go/border/rpkt/rpkt.go#L73)):*
> ```Go
> 	Ingress AddrIFPair
> 	// Egress is a list of function & address pairs that determine how and
> 	// where to the packet is sent. (PROCESS/ROUTE)
> ```

s/is sent/will be sent/

---

*[go/border/rpkt/rpkt.go, line 77 at r7](https://reviewable.io:443/reviews/netsec-ethz/scion/934#-KWs3kitI-FpmLRl5-Ms:-KWs3kitI-FpmLRl5-Mt:beluh27) ([raw file](https://github.com/netsec-ethz/scion/blob/e3dadbea96e3d342d8d916eb7d3354d175e94da2/go/border/rpkt/rpkt.go#L77)):*
> ```Go
> 	// CmnHdr is the SCION common header. Required for every packet. (PARSE)
> 	CmnHdr spkt.CmnHdr
> 	// idxs contains a set of indexs into Raw which point to the start of
> ```

indexs -> indexes or indices

---

*[go/border/rpkt/rpkt.go, line 100 at r7](https://reviewable.io:443/reviews/netsec-ethz/scion/934#-KWs3yLd9b1_iQ1qfRVg:-KWs3yLd9b1_iQ1qfRVh:bz989ey) ([raw file](https://github.com/netsec-ethz/scion/blob/e3dadbea96e3d342d8d916eb7d3354d175e94da2/go/border/rpkt/rpkt.go#L100)):*
> ```Go
> 	// HBHExt is the list of Hop-by-hop extensions, if any. (PARSE)
> 	HBHExt []RExtension
> 	// E2Ext is the list of end2end extensions, if any.
> ```

s/E2Ext/E2EExt/

---

*[go/border/rpkt/rpkt.go, line 112 at r7](https://reviewable.io:443/reviews/netsec-ethz/scion/934#-KWs46xoKRj2t1j9LRVO:-KWs46xoKRj2t1j9LRVP:b-jvlb8w) ([raw file](https://github.com/netsec-ethz/scion/blob/e3dadbea96e3d342d8d916eb7d3354d175e94da2/go/border/rpkt/rpkt.go#L112)):*
> ```Go
> 	pld common.Payload
> 	// hooks are registered callbacks to override/supplement normal processing.
> 	// The main use if for extensions to modify packet handling. (PARSE/PROCESS, only if needed)
> ```

The beginning of this sentence is broken.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion/934)*
<!-- Sent from Reviewable.io -->
